index.ts

```diff
@@ -0,0 +1,572 @@
+//      need to use `Object.defineProperty(Foo.prototype, ...)` to be able to make
+//      enumerable) to avoid object instantiation overhead and avoid creating objects for
+//      fields which are never read.
+//    - Map will need to be polyfilled or support for browsers before IE11 will need to
+//      be dropped.
+
+// ---------- core.ts ----------
+
+type ID = '' | string & {__isID: true};
+toID(s: string): ID;
+
+type primitive = string | number | boolean | undefined | null;
+type DeepReadonly<T> = T extends primitive ? T : DeepReadonlyObject<T>;
+type DeepReadonlyObject<T> = {
+  readonly [P in keyof T]: DeepReadonly<T[P]>
+}
+
+type Nullable<T> = {
+  [P in keyof T]: Nullable<T[P]> | null;
+}
+
+// ---------- conditions.ts ----------
+
+type WeatherName =
+  | 'Sand'
+  | 'Sun'
+  | 'Rain'
+  | 'Hail'
+  | 'Harsh Sunshine'
+  | 'Heavy Rain'
+  | 'Strong Winds';
+
+type Terrain =
+  | 'Psychic'
+  | 'Misty'
+  | 'Electric'
+  | 'Grassy';
+
+// TODO: volatiles/slot/side condition enums?
+
+type StatusName = 'PAR' | 'PSN' | 'FRZ' | 'SLP' | 'BRN' | 'TOX';
+
+// ---------- data.ts ----------
+
+type DataKind = 'Ability' | 'Move' | 'Item' | 'Species';
+type Nonstandard = 'Glitch' | 'Past' | 'Future' | 'CAP' | 'LGPE' | 'Pokestar' | 'Custom';
+
+interface JSONData extends Nullable<Omit<Data, 'id' | 'kind' | 'num' | 'gen'>> {}
+interface Data {
+  id: ID;
+  kind: DataKind;
+  name: string;
+  gen: Generation;
+  isUnreleased?: boolean;
+  isNonstandard?: Nonstandard;
+}
+
+// ---------- natures.ts ----------
+
+type NatureName = 'Adamant' | 'Bashful' | 'Bold' | 'Brave' | 'Calm' | 'Careful' | 'Docile' | 'Gentle' |
+  'Hardy' | 'Hasty' | 'Impish' | 'Jolly' | 'Lax' | 'Lonely' | 'Mild' | 'Modest' | 'Naive' | 'Naughty' |
+  'Quiet' | 'Quirky' | 'Rash' | 'Relaxed' | 'Sassy' | 'Serious' | 'Timid';
+
+interface Nature extends DeepReadonly<{
+  id: ID;
+  name: NatureName;
+  plus?: StatName;
+  minus?: StatName;
+
+  toJSON(): JSONNature;
+}> {}
+interface JSONNature = extends Nullable<Omit<Nature, 'id'>> {};
+
+interface Natures {
+  get(id: ID): Nature | undefined;
+  [Symbol.iterator](): IterableIterator<Nature>;
+}
+
+// ---------- abilities.ts ----------
+
+interface CommonAbility {
+  suppressesWeather?: boolean;
+  isUnsuppressible?: boolean;
+}
+
+interface JSONAbility extends Nullable<JSONData & CommonAbility> {}
+interface Ability extends DeepReadonly<Data & CommonAbility & {
+  kind: 'Ability';
+
+  toJSON(): JSONAbility;
+}> {}
+
+interface Abilities {
+  get(id: ID): Ability | undefined;
+  [Symbol.iterator](): IterableIterator<Ability>;
+}
+
+// ---------- items.ts ----------
+
+interface CommonItem<MoveT, SpeciesT, TypeT> {
+  fling?: {
+    basePower: number;
+    status?: StatusName;
+    volatileStatus?: ID;
+  };
+  forcedForme?: SpeciesT;
+  ignoreKlutz?: boolean;
+  infiltrates?: boolean;
+  isBerry?: boolean;
+  isChoice?: boolean;
+  isGem?: boolean;
+  isPokeball?: boolean;
+  megaEvolves?: SpeciesT;
+  megaStone?: SpeciesT;
```

Comment:
group, but i'm also wondering if there is some redundancy here? I think `megaEvolves` is the out-of-battle forme of `megaStone`